src/mlpack/bindings/go/mlpack/cli_util.go

```diff
@@ -0,0 +1,144 @@
+package mlpack
+
+/*
+#cgo CFLAGS: -I. -I/capi -g -Wall
+#cgo LDFLAGS: -L${SRCDIR} -Wl,-rpath,${SRCDIR} -lm -lmlpack -lgo_util
+#include <capi/cli_util.h>
+*/
+import "C"
+
+import (
+  "runtime"
+  "time"
+  "unsafe"
+)
+
+func hasParam(identifier string) bool {
+  return bool((C.mlpackHasParam(C.CString(identifier))))
+}
+
+func setPassed(identifier string) {
+  C.mlpackSetPassed(C.CString(identifier))
+}
+
+func setParamDouble(identifier string, value float64) {
+  C.mlpackSetParamDouble(C.CString(identifier), C.double(value))
+}
+
+func setParamInt(identifier string, value int) {
+  C.mlpackSetParamInt(C.CString(identifier), C.int(value))
+}
+func setParamFloat(identifier string, value float64) {
+  C.mlpackSetParamFloat(C.CString(identifier), C.float(value))
+}
+
+func setParamBool(identifier string, value bool) {
+  C.mlpackSetParamBool(C.CString(identifier), C.bool(value))
+}
+
+func setParamString(identifier string, value string) {
+  C.mlpackSetParamString(C.CString(identifier), C.CString(value))
+}
+
+func setParamPtr(identifier string, ptr unsafe.Pointer, copy bool) {
+  C.mlpackSetParamPtr(C.CString(identifier), (*C.double)(ptr), C.bool(copy))
+}
+func resetTimers() {
+  C.mlpackResetTimers()
+}
+
+func enableTimers() {
+  C.mlpackEnableTimers()
+}
+
+func disableBacktrace() {
+  C.mlpackDisableBacktrace()
+}
+
+func disableVerbose() {
+  C.mlpackDisableVerbose()
+}
+
+func enableVerbose() {
+  C.mlpackEnableVerbose()
+}
+
+func restoreSettings(method string) {
+  C.mlpackRestoreSettings(C.CString(method))
+}
+
+func clearSettings() {
+  C.mlpackClearSettings()
+}
+
+func getParamString(identifier string) string {
+  val := C.GoString(C.mlpackGetParamString(C.CString(identifier)))
+  return val
+}
+
+func getParamBool(identifier string) bool {
+  val := bool(C.mlpackGetParamBool(C.CString(identifier)))
+  return val
+}
+
+func getParamInt(identifier string) int {
+  val := int(C.mlpackGetParamInt(C.CString(identifier)))
+  return val
+}
+
+func getParamDouble(identifier string) float64 {
+  val := float64(C.mlpackGetParamDouble(C.CString(identifier)))
+  return val
+}
+
+type mlpackVectorType struct {
+  mem unsafe.Pointer
+}
+
+func (v *mlpackVectorType) allocVecIntPtr(identifier string) {
+  v.mem = C.mlpackGetVecIntPtr(C.CString(identifier))
+  runtime.KeepAlive(v)
+}
+
+func setParamVecInt(identifier string, vecInt []int) {
+  ptr := unsafe.Pointer(&vecInt[0])
+  C.mlpackSetParamVectorInt(C.CString(identifier), (*C.int64_t)(ptr),
```

Comment:
I have updated this part. Please take a look whether my approach is correct or not.

Reply:
Ok, so I did some digging and there is indeed some complexity here.  In Go, it looks like whether ints will be 4 or 8 bytes depends solely on the architecture: https://github.com/golang/go/blob/master/src/go/types/sizes.go#L158 (and also whether gccgo is being used, but I'm not too familiar).  In C or C++, though, this actually depends on the architecture!  So on 64-bit x86_64 Windows, `int` will be 4 bytes, but on 64-bit x86_64 Linux systems `int` is 8 bytes!

Since Go doesn't appear to choose sizes based on the underlying C/C++ sizes of `int`, I think we'll actually have to create a conversion layer here.

I took a look at the code changes that you made, and it looks like the idea is to pass only `long long`, which in basically every implementation out there is 64 bits.  (Technically that's not guaranteed, but, I think it will be a very long time until `long long` starts being 128 bits or something larger...)  This seems totally reasonable to me.

But I think there's one piece missing---what if `vecInt` is holding 32-bit integers?  Then when we do `ptr := unsafe.Pointer(&vecInt[0])`, we're actually going to pass a pointer to an array of 32-bit integers and treat it as though it's a pointer to an array of 64-bit integers.  So I think maybe we have to first convert `vecInt` from an `int[]` to an `int64[]`, or something like this.  Let me know what you think.

Also, it would be awesome if you could document the strategy here; i.e., make a note of why we're using `*C.longlong` as a type to pass.  This is pretty non-trivial, so I want to make sure that it makes sense to anyone who comes along to maintain this code. :)

Reply:
I have resolved the comments as suggested please take a look.

Reply:
Looks good to me.  If you like, it could be more efficient to also catch the case where `int` is already an `int64` (should be easy to check?).  In that situation we can avoid the copy and just use `vecInt` exactly as it is.